### PR TITLE
[12.0][ADD] crm_stage_multi_company

### PR DESCRIPTION
**Pull request by qgroulard**
This module adds support for multi company on crm stage.

### CODE REVIEW

File: crm_stage_multi_company/__init__.py [hunk 1]
```diff
@@ -0,0 +1 @@
+from . import models
```
Comment (rousseldenis) [inline crm_stage_multi_company/__init__.py]: Missing line

File: crm_stage_multi_company/__manifest__.py [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+{
+    "name": "Crm Stage Multi Company",
+    "summary": """
+        This module adds support for multi company on crm stage.""",
+    "version": "12.0.1.0.0",
+    "license": "AGPL-3",
+    "author": "ACS0NE SA/NV, Odoo Community Association (OCA)",
+    "website": "https://acsone.eu",
```
Comment (rousseldenis) [inline crm_stage_multi_company/__manifest__.py]: Could you follow OCA conventions like in : https://github.com/OCA/maintainer-tools/blob/master/template/module/__manifest__.py

by adding missing entries and modifying website ?

File: crm_stage_multi_company/models/crm_stage.py [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+    _inherit = "crm.stage"
+
+    company_id = fields.Many2one(
+        "res.company",
+        string="Company",
+        index=True,
+        help="Specific company that uses this stage. "
+        "Other companies will not be able to see or use this stage.",
+        domain=lambda self: [("id", "child_of", self.env.user.company_id.id)],
+        default=lambda self: self.env.user.company_id.id,
```
Comment (rousseldenis) [inline crm_stage_multi_company/models/crm_stage.py]: Maybe a function to allow override ?

**Review by rousseldenis — CHANGES_REQUESTED** [review]
Code review. LGTM.

Some comments

**Comment by rousseldenis** [comment]
@qgroulard Could you please check travis ?

File: crm_stage_multi_company/models/crm_stage.py [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+        return [("id", "child_of", self.env.user.company_id.id)]
+
+    company_id = fields.Many2one(
+        "res.company",
+        string="Company",
+        index=True,
+        help="Specific company that uses this stage. "
+        "Other companies will not be able to see or use this stage.",
+        domain=lambda self: self._get_company_id_domain(),
+        default=lambda self: self._get_default_company_id(),
```
Comment (sbidoul) [inline crm_stage_multi_company/models/crm_stage.py]: @qgroulard the use pattern is tu use company_default_get so:

```suggestion
        default=lambda self: self.env['res.company']._company_default_get('crm.stage'),
```

File: crm_stage_multi_company/security/crm_security.xml [hunk 5]
```diff
@@ -0,0 +1,10 @@
+<?xml version="1.0" encoding="utf-8"?>
+<odoo>
+
+    <record id="crm_stage_company_rule" model="ir.rule">
+        <field name="name">CRM Stage Multi-Company</field>
+        <field name="model_id" ref="model_crm_stage"/>
```
Comment (sbidoul) [inline crm_stage_multi_company/security/crm_security.xml]: Don't we need the module prefix in ref?

File: crm_stage_multi_company/models/crm_stage.py [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+    def _get_company_id_domain(self):
+        return [("id", "child_of", self.env.user.company_id.id)]
+
+    company_id = fields.Many2one(
+        "res.company",
+        string="Company",
+        index=True,
+        help="Specific company that uses this stage. "
+        "Other companies will not be able to see or use this stage.",
+        domain=lambda self: self._get_company_id_domain(),
```
Comment (sbidoul) [inline crm_stage_multi_company/models/crm_stage.py]: I don't think a domain is very useful here.

File: crm_stage_multi_company/models/crm_stage.py [hunk 7]
```diff
@@ -8,18 +8,14 @@ class CrmStage(models.Model):
 
     _inherit = "crm.stage"
 
-    def _get_default_company_id(self):
+    def _company_default_get(self):
         return self.env.user.company_id.id
```
Comment (sbidoul) [inline crm_stage_multi_company/models/crm_stage.py]: @qgroulard this method is useless, right?

**Review by sbejaoui — APPROVED** [review]
Thank you @qgroulard (y)

**Review by RobinetDenisAcsone — APPROVED** [review]
LGTM

**Comment by sbejaoui** [comment]
@rousseldenis , can you update you review please

**Review by rousseldenis — APPROVED** [review]
Code review.

**Comment by rousseldenis** [comment]
@OCA/core-maintainers

**Comment by moylop260** [comment]
<s>Sorry, I have aborted the command to merge...

There are still comments</s>

Forget it
They are outdated...

**Comment by moylop260** [comment]
Could you using another name of company?

There is a issue merging because of it is duplicated:
 - https://travis-ci.org/OCA/multi-company/jobs/595383509#L1582

**Pull request merged**